File: apis/pkg.py
from collections.abc import Generator, Iterator
from dataclasses import dataclass, field
from pathlib import Path
import os
import shutil
import stat
import subprocess
import tomllib

from pyrsult import Result, Success, Failure

from ._error import Error, IOError, GitClone, GitPull, NotFound
from ._event import Event, Progress, Log
from .types import Config, parse_pack_name
from .vindex import build_dep_tree, get_transitive_deps


@dataclass
class PackageInfo:
    host: str
    user: str
    repo: str
    full_name: str
    path: Path
    has_vindex: bool


@dataclass
class PruneReport:
    removed_invalid: list[str] = field(default_factory=list)
    removed_empty: list[str] = field(default_factory=list)
    removed_unused: list[str] = field(default_factory=list)


@dataclass
class UpdateInfo:
    full_name: str
    updated: bool


def _remove_readonly(func, path, exc_info):
    os.chmod(path, stat.S_IWRITE)
    func(path)


def iter_package_dirs(libs_path: Path) -> Iterator[tuple[Path, Path, Path, str]]:
    for md in libs_path.iterdir():
        if not md.is_dir():
            continue
        for ud in md.iterdir():
            if not ud.is_dir():
                continue
            for rd in ud.iterdir():
                if not rd.is_dir():
                    continue
                yield md, ud, rd, f"{md.name}:{ud.name}.{rd.name}"


def iter_empty_dirs(libs_path: Path) -> Iterator[Path]:
    for md in sorted(libs_path.iterdir(), reverse=True):
        if not md.is_dir():
            continue
        for ud in sorted(md.iterdir(), reverse=True):
            if not ud.is_dir():
                continue
            for rd in sorted(ud.iterdir(), reverse=True):
                if not rd.is_dir():
                    continue
                if not any(rd.iterdir()):
                    yield rd
            if not any(ud.iterdir()):
                yield ud
        if not any(md.iterdir()):
            yield md


def _git_clone(
    url: str, dest: Path, branch: str | None = None
) -> Result[None, GitClone]:
    cmd = ["git", "clone", url, str(dest)]
    if branch:
        cmd.extend(["-b", branch])
    try:
        result = subprocess.run(cmd, capture_output=True, text=True)
        if result.returncode != 0:
            return Failure(GitClone(url=url, detail=result.stderr.strip()))
        return Success(None)
    except FileNotFoundError:
        return Failure(GitClone(url=url, detail="未找到 git 命令"))
    except OSError as e:
        return Failure(GitClone(url=url, detail=str(e)))


def _git_pull(path: Path) -> Result[bool, GitPull]:
    try:
        result = subprocess.run(
            ["git", "-C", str(path), "pull"], capture_output=True, text=True
        )
        if result.returncode != 0:
            return Failure(GitPull(path=str(path), detail=result.stderr.strip()))
        already_up_to_date = (
            "Already up to date" in result.stdout or "已经是最新的" in result.stdout
        )
        return Success(not already_up_to_date)
    except FileNotFoundError:
        return Failure(GitPull(path=str(path), detail="未找到 git 命令"))
    except OSError as e:
        return Failure(GitPull(path=str(path), detail=str(e)))


def list_packages() -> Result[list[PackageInfo], Error]:
    libs_path = Config.local_libs_path()
    if not libs_path.exists():
        return Failure(NotFound("libs_path", str(libs_path)))

    packages: list[PackageInfo] = []
    for _, user_dir, repo_dir, full_name in iter_package_dirs(libs_path):
        vindex_file = repo_dir / "vindex.toml"
        packages.append(
            PackageInfo(
                host=repo_dir.parent.parent.name,
                user=user_dir.name,
                repo=repo_dir.name,
                full_name=full_name,
                path=repo_dir,
                has_vindex=vindex_file.exists(),
            )
        )
    return Success(packages)


def delete_package(spec: str) -> Result[None, Error]:
    paths_to_check = [Config.local_libs_path(), Config.VIX_LIBS_PATH]
    for libs_path in paths_to_check:
        info = parse_pack_name(spec, parent=libs_path)
        pack_path = info.pack_path
        if pack_path.exists():
            try:
                shutil.rmtree(pack_path, onexc=_remove_readonly)
            except OSError as e:
                return Failure(IOError(path=str(pack_path), detail=str(e)))
            return Success(None)
    return Failure(NotFound("package", spec))


def prune_packages(
    empty_only: bool = False,
    invalid_only: bool = False,
    remove_unused: bool = False,
) -> Result[PruneReport, Error]:
    libs_path = Config.local_libs_path()
    if not libs_path.exists():
        return Failure(NotFound("libs_path", str(libs_path)))

    report = PruneReport()

    if not empty_only:
        for _, _, repo_dir, full_name in iter_package_dirs(libs_path):
            has_vindex = (repo_dir / "vindex.toml").exists()
            if not has_vindex:
                report.removed_invalid.append(full_name)
                shutil.rmtree(repo_dir, onexc=_remove_readonly)

    if not invalid_only:
        for empty_dir in iter_empty_dirs(libs_path):
            rel = str(empty_dir.relative_to(libs_path))
            report.removed_empty.append(rel)
            empty_dir.rmdir()

    should_find_unused = remove_unused or (not empty_only and not invalid_only)
    if should_find_unused:
        unused = _find_unused_packages(libs_path)
        for _, _, repo_dir, full_name in iter_package_dirs(libs_path):
            if full_name in unused:
                report.removed_unused.append(full_name)
                shutil.rmtree(repo_dir, onexc=_remove_readonly)

    return Success(report)


def _find_unused_packages(libs_path: Path) -> list[str]:
    vindex_path = Path.cwd() / "vindex.toml"
    root_deps: list[str] = []
    if vindex_path.exists():
        with open(vindex_path, "rb") as f:
            data = tomllib.load(f)
        root_deps = data.get("project", {}).get("deps", [])
        legacy = list(data.get("dependencies", {}).keys())
        root_deps = list(dict.fromkeys(root_deps + legacy))

    referenced = build_dep_tree(libs_path, root_deps)
    unused: list[str] = []
    for _, _, _, full_name in iter_package_dirs(libs_path):
        if full_name not in referenced:
            unused.append(full_name)
    return unused


def install_package(
    spec: str, force_local: bool = False
) -> Generator[Event, None, Result[PackageInfo, Error]]:
    yield Log("info", f"正在安装 {spec}")

    libs_path = Config.local_libs_path()
    global_path = Config.VIX_LIBS_PATH

    info = parse_pack_name(spec, parent=libs_path)
    dest = info.pack_path

    if dest.exists():
        yield Log("info", f"包已存在: {info.full_name}")
        return Success(
            PackageInfo(
                host=info.git_master,
                user=info.user_name,
                repo=info.repo_name,
                full_name=info.full_name,
                path=dest,
                has_vindex=(dest / "vindex.toml").exists(),
            )
        )

    if not force_local:
        global_info = parse_pack_name(spec, parent=global_path)
        global_dest = global_info.pack_path
        if global_dest.exists():
            yield Log("info", f"包已存在于全局: {global_info.full_name}")
            return Success(
                PackageInfo(
                    host=global_info.git_master,
                    user=global_info.user_name,
                    repo=global_info.repo_name,
                    full_name=global_info.full_name,
                    path=global_dest,
                    has_vindex=(global_dest / "vindex.toml").exists(),
                )
            )

    libs_path.mkdir(parents=True, exist_ok=True)
    (libs_path / info.git_master).mkdir(parents=True, exist_ok=True)
    (libs_path / info.git_master / info.user_name).mkdir(parents=True, exist_ok=True)

    yield Progress(f"克隆 {info.git_url}")

    clone_result = _git_clone(info.git_url, dest, info.branch_name)
    if isinstance(clone_result, Failure):
        yield Log("error", clone_result.error.detail)
        return clone_result

    yield Log("info", "克隆完成")

    vindex_file = dest / "vindex.toml"
    has_vindex = vindex_file.exists()

    if not has_vindex:
        yield Log("warn", "包缺少 vindex.toml")

    pack_info = PackageInfo(
        host=info.git_master,
        user=info.user_name,
        repo=info.repo_name,
        full_name=info.full_name,
        path=dest,
        has_vindex=has_vindex,
    )

    if has_vindex:
        deps = get_transitive_deps(dest)
        if deps:
            yield Log("info", f"安装 {len(deps)} 个依赖...")
            for dep_spec in deps:
                yield Progress(f"安装依赖 {dep_spec}")
                dep_result = yield from install_package(
                    dep_spec, force_local=force_local
                )
                if isinstance(dep_result, Failure):
                    yield Log("error", f"依赖 {dep_spec} 安装失败: {dep_result.error}")

    return Success(pack_info)


def update_package(spec: str) -> Generator[Event, None, Result[UpdateInfo, Error]]:
    yield Log("info", f"正在更新 {spec}")

    paths_to_check = [Config.local_libs_path(), Config.VIX_LIBS_PATH]
    pack_path = None
    full_name = None

    for libs_path in paths_to_check:
        info = parse_pack_name(spec, parent=libs_path)
        if info.pack_path.exists():
            pack_path = info.pack_path
            full_name = info.full_name
            break

    if pack_path is None:
        return Failure(NotFound("package", spec))

    assert full_name is not None
    yield Progress(f"拉取 {full_name}")
    pull_result = _git_pull(pack_path)
    if isinstance(pull_result, Failure):
        return pull_result

    updated = pull_result.unwrap()
    if updated:
        yield Log("info", f"已更新 {full_name}")
    else:
        yield Log("info", f"{full_name} 已是最新")

    vindex_file = pack_path / "vindex.toml"
    if vindex_file.exists():
        deps = get_transitive_deps(pack_path)
        for dep_spec in deps:
            yield from update_package(dep_spec)

    return Success(UpdateInfo(full_name=full_name, updated=updated))


def install_dependencies(
    force_local: bool = False,
) -> Generator[Event, None, Result[list[PackageInfo], Error]]:
    vindex_path = Path.cwd() / "vindex.toml"
    if not vindex_path.exists():
        return Failure(NotFound("vindex.toml", str(vindex_path)))

    with open(vindex_path, "rb") as f:
        data = tomllib.load(f)

    dep_specs = data.get("project", {}).get("deps", [])
    legacy = list(data.get("dependencies", {}).keys())
    dep_specs = list(dict.fromkeys(dep_specs + legacy))

    if not dep_specs:
        yield Log("info", "没有依赖需要安装")
        return Success([])

    installed: list[PackageInfo] = []
    for dep_spec in dep_specs:
        yield Log("info", f"检查依赖: {dep_spec}")
        result = yield from install_package(dep_spec, force_local=force_local)
        if isinstance(result, Success):
            installed.append(result.unwrap())
        else:
            yield Log("error", f"依赖 {dep_spec} 安装失败: {result.error}")

    return Success(installed)

File: internal/api/pkg.go
package api

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"

	"github.com/BurntSushi/toml"
)

type PackageInfo struct {
	Host      string
	User      string
	Repo      string
	FullName  string
	Path      string
	HasVIndex bool
}

type PruneReport struct {
	RemovedInvalid []string
	RemovedEmpty   []string
	RemovedUnused  []string
}

type UpdateInfo struct {
	FullName string
	Updated  bool
}

// iterPackageDirs 遍历 libs 目录下的 host/user/repo 三级嵌套目录。
// 返回 Go 1.23 风格的迭代器，供 ListPackages/PrunePackages 等函数消费。
func iterPackageDirs(libsPath string) func(func(string, string, string, string) bool) {
	return func(yield func(host, user, repo, fullName string) bool) {
		entries, err := os.ReadDir(libsPath)
		if err != nil {
			return
		}
		for _, hostEntry := range entries {
			if !hostEntry.IsDir() {
				continue
			}
			hostDir := filepath.Join(libsPath, hostEntry.Name())
			userEntries, err := os.ReadDir(hostDir)
			if err != nil {
				continue
			}
			for _, userEntry := range userEntries {
				if !userEntry.IsDir() {
					continue
				}
				userDir := filepath.Join(hostDir, userEntry.Name())
				repoEntries, err := os.ReadDir(userDir)
				if err != nil {
					continue
				}
				for _, repoEntry := range repoEntries {
					if !repoEntry.IsDir() {
						continue
					}
					fullName := fmt.Sprintf("%s:%s.%s", hostEntry.Name(), userEntry.Name(), repoEntry.Name())
					if !yield(hostEntry.Name(), userEntry.Name(), repoEntry.Name(), fullName) {
						return
					}
				}
			}
		}
	}
}

func fileExists(path string) bool {
	_, err := os.Stat(path)
	return err == nil
}

func gitClone(url, dest, branch string) error {
	args := []string{"clone", url, dest}
	if branch != "" {
		args = append(args, "-b", branch)
	}
	cmd := exec.Command("git", args...)
	out, err := cmd.CombinedOutput()
	if err != nil {
		return &GitClone{URL: url, Detail: strings.TrimSpace(string(out))}
	}
	return nil
}

func gitPull(path string) (bool, error) {
	cmd := exec.Command("git", "-C", path, "pull")
	out, err := cmd.CombinedOutput()
	if err != nil {
		return false, &GitPull{Path: path, Detail: strings.TrimSpace(string(out))}
	}
	output := string(out)
	alreadyUpToDate := strings.Contains(output, "Already up to date") || strings.Contains(output, "已经是最新的")
	return !alreadyUpToDate, nil
}

func removeReadonly(path string) {
	os.RemoveAll(path)
}

func ListPackages() ([]*PackageInfo, error) {
	libsPath := Config{}.LocalLibsPath()
	if !fileExists(libsPath) {
		return nil, &NotFound{Kind: "libs_path", Name: libsPath}
	}

	var packages []*PackageInfo
	iterPackageDirs(libsPath)(func(host, user, repo, fullName string) bool {
		repoDir := filepath.Join(libsPath, host, user, repo)
		packages = append(packages, &PackageInfo{
			Host:      host,
			User:      user,
			Repo:      repo,
			FullName:  fullName,
			Path:      repoDir,
			HasVIndex: fileExists(filepath.Join(repoDir, "vindex.toml")),
		})
		return true
	})
	return packages, nil
}

func DeletePackage(spec string) error {
	pathsToCheck := []string{Config{}.LocalLibsPath(), Config{}.VIX_LIBS_PATH()}
	for _, libsPath := range pathsToCheck {
		info, err := ParsePackName(spec, libsPath)
		if err != nil {
			continue
		}
		packPath := info.PackPath()
		if fileExists(packPath) {
			if err := os.RemoveAll(packPath); err != nil {
				return &IOError{Path: packPath, Detail: err.Error()}
			}
			return nil
		}
	}
	return &NotFound{Kind: "package", Name: spec}
}

func FindUnusedPackages(libsPath string) map[string]bool {
	vindexPath := filepath.Join(".", "vindex.toml")
	var rootDeps []string
	if data, err := os.ReadFile(vindexPath); err == nil {
		var m map[string]any
		if err := toml.Unmarshal(data, &m); err == nil {
			if proj, ok := m["project"].(map[string]any); ok {
				if deps, ok := proj["deps"].([]any); ok {
					for _, d := range deps {
						if s, ok := d.(string); ok {
							rootDeps = append(rootDeps, s)
						}
					}
				}
			}
			if deps, ok := m["dependencies"].(map[string]any); ok {
				for d := range deps {
					rootDeps = append(rootDeps, d)
				}
			}
		}
	}

	referenced := BuildDepTree(libsPath, rootDeps)
	unused := make(map[string]bool)
	iterPackageDirs(libsPath)(func(host, user, repo, fullName string) bool {
		if !referenced[fullName] {
			unused[fullName] = true
		}
		return true
	})
	return unused
}

func PrunePackages(emptyOnly, invalidOnly, removeUnused bool) (*PruneReport, error) {
	libsPath := Config{}.LocalLibsPath()
	if !fileExists(libsPath) {
		return nil, &NotFound{Kind: "libs_path", Name: libsPath}
	}

	report := &PruneReport{}
	SpecificFlag := emptyOnly || invalidOnly || removeUnused

	if !emptyOnly && !(removeUnused && !invalidOnly) {
		iterPackageDirs(libsPath)(func(host, user, repo, fullName string) bool {
			repoDir := filepath.Join(libsPath, host, user, repo)
			if !fileExists(filepath.Join(repoDir, "vindex.toml")) {
				report.RemovedInvalid = append(report.RemovedInvalid, fullName)
				os.RemoveAll(repoDir)
			}
			return true
		})
	}

	if !invalidOnly && !(removeUnused && !emptyOnly) {
		removeEmptyDirs(libsPath, &report.RemovedEmpty)
	}

	shouldFindUnused := removeUnused || (!SpecificFlag)
	if shouldFindUnused {
		unused := FindUnusedPackages(libsPath)
		iterPackageDirs(libsPath)(func(host, user, repo, fullName string) bool {
			if unused[fullName] {
				report.RemovedUnused = append(report.RemovedUnused, fullName)
				repoDir := filepath.Join(libsPath, host, user, repo)
				os.RemoveAll(repoDir)
			}
			return true
		})
	}

	return report, nil
}

// removeEmptyDirs 从最深层向根目录删除空目录，避免父目录非空导致无法删除子目录。
func removeEmptyDirs(libsPath string, removed *[]string) {
	entries, _ := os.ReadDir(libsPath)
	for i := len(entries) - 1; i >= 0; i-- {
		hostDir := filepath.Join(libsPath, entries[i].Name())
		userEntries, _ := os.ReadDir(hostDir)
		for j := len(userEntries) - 1; j >= 0; j-- {
			userDir := filepath.Join(hostDir, userEntries[j].Name())
			repoEntries, _ := os.ReadDir(userDir)
			for k := len(repoEntries) - 1; k >= 0; k-- {
				repoDir := filepath.Join(userDir, repoEntries[k].Name())
				if isEmptyDir(repoDir) {
					*removed = append(*removed, relPath(libsPath, repoDir))
					os.Remove(repoDir)
				}
			}
			if isEmptyDir(userDir) {
				*removed = append(*removed, relPath(libsPath, userDir))
				os.Remove(userDir)
			}
		}
		if isEmptyDir(hostDir) {
			*removed = append(*removed, relPath(libsPath, hostDir))
			os.Remove(hostDir)
		}
	}
}

func isEmptyDir(path string) bool {
	entries, err := os.ReadDir(path)
	if err != nil {
		return false
	}
	return len(entries) == 0
}

func relPath(base, target string) string {
	rel, err := filepath.Rel(base, target)
	if err != nil {
		return target
	}
	return rel
}

func InstallPackage(spec string, forceLocal bool) (*PackageInfo, error) {
	libsPath := Config{}.LocalLibsPath()
	globalPath := Config{}.VIX_LIBS_PATH()

	info, err := ParsePackName(spec, libsPath)
	if err != nil {
		return nil, err
	}
	dest := info.PackPath()

	if fileExists(dest) {
		return &PackageInfo{
			Host:      info.GitMaster,
			User:      info.UserName,
			Repo:      info.RepoName,
			FullName:  info.FullName(),
			Path:      dest,
			HasVIndex: fileExists(filepath.Join(dest, "vindex.toml")),
		}, nil
	}

	if !forceLocal {
		globalInfo, err := ParsePackName(spec, globalPath)
		if err == nil {
			globalDest := globalInfo.PackPath()
			if fileExists(globalDest) {
				return &PackageInfo{
					Host:      globalInfo.GitMaster,
					User:      globalInfo.UserName,
					Repo:      globalInfo.RepoName,
					FullName:  globalInfo.FullName(),
					Path:      globalDest,
					HasVIndex: fileExists(filepath.Join(globalDest, "vindex.toml")),
				}, nil
			}
		}
	}

	if err := os.MkdirAll(libsPath, 0755); err != nil {
		return nil, &IOError{Path: libsPath, Detail: err.Error()}
	}
	hostDir := filepath.Join(libsPath, info.GitMaster)
	userDir := filepath.Join(hostDir, info.UserName)
	if err := os.MkdirAll(userDir, 0755); err != nil {
		return nil, &IOError{Path: userDir, Detail: err.Error()}
	}

	if err := gitClone(info.GitURL(), dest, info.BranchName); err != nil {
		return nil, err
	}

	hasVIndex := fileExists(filepath.Join(dest, "vindex.toml"))

	packInfo := &PackageInfo{
		Host:      info.GitMaster,
		User:      info.UserName,
		Repo:      info.RepoName,
		FullName:  info.FullName(),
		Path:      dest,
		HasVIndex: hasVIndex,
	}

	if hasVIndex {
		deps := GetTransitiveDeps(dest)
		for _, depSpec := range deps {
			_, err := InstallPackage(depSpec, forceLocal)
			if err != nil {
				_ = err // log in CLI layer
			}
		}
	}

	return packInfo, nil
}

func UpdatePackage(spec string) (*UpdateInfo, error) {
	return updatePackageInternal(spec, make(map[string]bool))
}

func updatePackageInternal(spec string, visited map[string]bool) (*UpdateInfo, error) {
	if visited[spec] {
		return &UpdateInfo{FullName: spec, Updated: false}, nil
	}
	visited[spec] = true

	pathsToCheck := []string{Config{}.LocalLibsPath(), Config{}.VIX_LIBS_PATH()}
	var packPath string
	var fullName string

	for _, libsPath := range pathsToCheck {
		info, err := ParsePackName(spec, libsPath)
		if err != nil {
			continue
		}
		if fileExists(info.PackPath()) {
			packPath = info.PackPath()
			fullName = info.FullName()
			break
		}
	}

	if packPath == "" {
		return nil, &NotFound{Kind: "package", Name: spec}
	}

	updated, err := gitPull(packPath)
	if err != nil {
		return nil, err
	}

	vindexFile := filepath.Join(packPath, "vindex.toml")
	if fileExists(vindexFile) {
		deps := GetTransitiveDeps(packPath)
		for _, depSpec := range deps {
			_, _ = updatePackageInternal(depSpec, visited)
		}
	}

	return &UpdateInfo{FullName: fullName, Updated: updated}, nil
}

func InstallDependencies(forceLocal bool) ([]*PackageInfo, error) {
	vindexPath := filepath.Join(".", "vindex.toml")
	if !fileExists(vindexPath) {
		return nil, &NotFound{Kind: "vindex.toml", Name: vindexPath}
	}

	data, err := os.ReadFile(vindexPath)
	if err != nil {
		return nil, &IOError{Path: vindexPath, Detail: err.Error()}
	}

	var m map[string]any
	if err := toml.Unmarshal(data, &m); err != nil {
		return nil, &IOError{Path: vindexPath, Detail: err.Error()}
	}

	var depSpecs []string
	if proj, ok := m["project"].(map[string]any); ok {
		if deps, ok := proj["deps"].([]any); ok {
			for _, d := range deps {
				if s, ok := d.(string); ok {
					depSpecs = append(depSpecs, s)
				}
			}
		}
	}
	if deps, ok := m["dependencies"].(map[string]any); ok {
		for d := range deps {
			depSpecs = append(depSpecs, d)
		}
	}

	deduped := make([]string, 0, len(depSpecs))
	seen := make(map[string]bool)
	for _, d := range depSpecs {
		if !seen[d] {
			seen[d] = true
			deduped = append(deduped, d)
		}
	}

	if len(deduped) == 0 {
		return []*PackageInfo{}, nil
	}

	var installed []*PackageInfo
	for _, depSpec := range deduped {
		packInfo, err := InstallPackage(depSpec, forceLocal)
		if err != nil {
			continue
		}
		installed = append(installed, packInfo)
	}

	return installed, nil
}

func ReadPackageReadme(spec string) (string, error) {
	pathsToCheck := []string{Config{}.LocalLibsPath(), Config{}.VIX_LIBS_PATH()}
	var packPath string

	for _, libsPath := range pathsToCheck {
		info, err := ParsePackName(spec, libsPath)
		if err != nil {
			continue
		}
		if fileExists(info.PackPath()) {
			packPath = info.PackPath()
			break
		}
	}

	if packPath == "" {
		return "", &NotFound{Kind: "package", Name: spec}
	}

	readmeName := "README.md"
	vindexFile := filepath.Join(packPath, "vindex.toml")
	if data, err := os.ReadFile(vindexFile); err == nil {
		var m map[string]any
		if err := toml.Unmarshal(data, &m); err == nil {
			if proj, ok := m["project"].(map[string]any); ok {
				if r, ok := proj["readme"].(string); ok {
					readmeName = r
				}
			}
		}
	}

	readmePath := filepath.Join(packPath, readmeName)
	if !fileExists(readmePath) {
		return "", &NotFound{Kind: "readme", Name: readmePath}
	}

	data, err := os.ReadFile(readmePath)
	if err != nil {
		return "", &IOError{Path: readmePath, Detail: err.Error()}
	}

	return string(data), nil
}
